velox/dwio/parquet/writer/Writer.cpp

```diff
@@ -138,49 +143,101 @@ Writer::Writer(
               "writer_node_{}",
               folly::to<std::string>(folly::Random::rand64())))} {}
 
+void Writer::flush() {
+  if (stagingRows_ > 0) {
+    if (!arrowContext_->writer) {
+      auto arrowProperties =
+          ::parquet::ArrowWriterProperties::Builder().build();
+      PARQUET_THROW_NOT_OK(::parquet::arrow::FileWriter::Open(
+          *arrowContext_->schema.get(),
+          arrow::default_memory_pool(),
+          stream_,
+          arrowContext_->properties,
+          arrowProperties,
+          &arrowContext_->writer));
+    }
+
+    auto fields = arrowContext_->schema->fields();
+    std::vector<std::shared_ptr<arrow::ChunkedArray>> chunks;
+    for (int colIdx = 0; colIdx < fields.size(); colIdx++) {
+      auto dataType = fields.at(colIdx)->type();
+      auto chunk =
+          arrow::ChunkedArray::Make(
+              std::move(arrowContext_->stagingChunks.at(colIdx)), dataType)
+              .ValueOrDie();
+      chunks.push_back(chunk);
+    }
+    auto table = arrow::Table::Make(
+        arrowContext_->schema, std::move(chunks), stagingRows_);
+    PARQUET_THROW_NOT_OK(
+        arrowContext_->writer->WriteTable(*table, rowsInRowGroup_));
+    PARQUET_THROW_NOT_OK(stream_->Flush());
+    for (auto& chunk : arrowContext_->stagingChunks) {
+      chunk.clear();
+    }
+    stagingRows_ = 0;
+    stagingBytes_ = 0;
+  }
+}
+
+/**
+ * This method would cache input `ColumnarBatch` to make the size of row group
+ * big. It would flush when:
+ * - the cached numRows bigger than `rowsInRowGroup_`
+ * - the cached bytes bigger than `bytesInRowGroup_`
+ *
+ * This method assumes each input `ColumnarBatch` have same schema.
+ */
 void Writer::write(const VectorPtr& data) {
   ArrowArray array;
   ArrowSchema schema;
   exportToArrow(data, array, generalPool_.get());
   exportToArrow(data, schema);
   PARQUET_ASSIGN_OR_THROW(
       auto recordBatch, arrow::ImportRecordBatch(&array, &schema));
-  auto table = arrow::Table::Make(
-      recordBatch->schema(), recordBatch->columns(), data->size());
-  if (!arrowContext_->writer) {
-    auto arrowProperties = ::parquet::ArrowWriterProperties::Builder().build();
-    PARQUET_THROW_NOT_OK(::parquet::arrow::FileWriter::Open(
-        *recordBatch->schema(),
-        arrow::default_memory_pool(),
-        stream_,
-        arrowContext_->properties,
-        arrowProperties,
-        &arrowContext_->writer));
+  if (!arrowContext_->schema) {
+    arrowContext_->schema = recordBatch->schema();
+    for (int colIdx = 0; colIdx < arrowContext_->schema->num_fields();
+         colIdx++) {
+      arrowContext_->stagingChunks.push_back(
+          std::vector<std::shared_ptr<arrow::Array>>());
+    }
+  }
+
+  auto bytes = data->estimateFlatSize();
+  auto numRows = data->size();
+  if (stagingBytes_ + bytes > bytesInRowGroup_ ||
+      stagingRows_ + numRows > rowsInRowGroup_) {
+    flush();
```

Comment:
@JkSelf Here we are changing to automatic flush for row groups.  Is it ok with Gluten to do that?

Reply:
I think it's ok to Gluten. Gluten only call `close` at the end of write task, so the behavior of `flush` does not affect Gluten.

Reply:
Yes. It's ok to Gluten. This can resolve the small row group issue in Gluten.